Add session that blows up when using internal API

dstandish opened the pull request:
Here I add a TracebackSession which is designed to blow up the first time it's used and then provide a traceback for the original call site.  I also ensure that this is the Session class that is used when internal API config is enabled.

dstandish committed:
Make _run_task_by_local_task_job compatible with internal API

uranusjr commented on airflow/cli/commands/task_command.py:
```diff
@@ -278,7 +279,15 @@ def _run_task_by_local_task_job(args, ti: TaskInstance | TaskInstancePydantic) -
+        # by the provide_session decorator.  All downstream functions that actually use
+        # a session are already set up for RPC
+        # todo: perhaps instead we should simply modify the provide_session decorator
+        # to *not* provide a session when using internal API
```
Thinking the same, the decorator should just be made smarter about this.

dstandish replied:
so to be clear, you think `@provide_session` should just provide `None` object when internal API configured?  @potiuk thoughts?

dstandish replied:
@uranusjr thoughts?

potiuk replied:
I'd say provide_session should provide a fake session that should:

1) store traceback where it has  been provided for the first time
2) raise an exception when it is accessed for anything explaining both:

  a) where it has been provided
  b) where it has been used
   c) eventually explanation what to do to contributor. 

Both (in CI / breeze / development environment) using `rich`'s traceback functionality (https://rich.readthedocs.io/en/stable/traceback.html).

I see this as CI/Development/Diagnostic  tool. If we return None, the most we will get is an exception that None has missing property - which willl be difficult to really trace, because the session can be passed throuh multiple layers and multiple method calls before it is used. 

When we enable internal API, there will be a test in CI failing if somone will add a new DB call and did not test it with internal API, so what we really want in this case is to provide the fastest path for the contributor to find out what they should do to fix the problem. I don't think passing None is good enough for that.

dstandish replied:
i'm working on something like this

dstandish committed:
Add session that blows up when using internal API

Here I add a TracebackSession which is designed to blow up the first time it's used and then provide a traceback for the original call site.  I also ensure that this is the Session class that is used when internal API config is enabled.

dstandish committed:
remove custom session avoidance logic

dstandish committed:
revert some other changes

potiuk commented on airflow/utils/session.py:
```diff
@@ -24,12 +24,17 @@
 @contextlib.contextmanager
 def create_session() -> Generator[SASession, None, None]:
     """Contextmanager that will create and teardown a session."""
+    if InternalApiConfig.get_use_internal_api():
```
should be:

```
if conf.getboolean("core", "database_access_isolation", fallback=False)
```

dstandish replied:
Why?
I thought that `database_access_isolation` might be true globally for the deployment but e.g. scheduler and webserver would disregard that and still connect to DB?  I.e. with `force_database_direct_access()`?

If that's not true, perhaps we should remove `force_database_direct_access` and just throw an error if scheduler / webserver have `database_access_isolation` True?

potiuk replied:
Ah yes. Now I re-read it, it's been a while.

dstandish committed:
dont force db isolation in unit tests
